### DISCUSSION TITLE
Using sensors without starting `hub_runtime`

### DISCUSSION THREAD

**Discussion by aivarannamaa** [gpdaniels/spike-prime#35] [General]
I'm looking for ways to use the official sensors API in `spike` / `mindstorms` modules without importing and starting the heavy `hub_runtime`.

The first obstacle was importing `mindstorms` at clean slate, because the transitive imports exceed the allowed stack height. The work-around was to import another module (eg. `_api` or `system`) first:

```python
import system
from mindstorms import MSHub, Motor, ColorSensor
```

I can now get a working instance of `Motor`:

```python
m = Motor("A")
m.run_for_seconds(1)
```
... but not a `ColorSensor` (or `DistanceSensor`).

`cs = ColorSensor("E")` gives me:

```
Traceback (most recent call last):
  File "<stdin>", line 7, in <module>
  File "_api/colorsensor.py", line 1, in __init__
RuntimeError: There is not a color sensor connected to port E
```

although the device is visible via `hub` API (`hub.port.E.device`)

If I add

```python
import hub_runtime
hub_runtime.init(0) # don't know what the argument means
```

then I can construct the ColorsSensor without problems, but I'd like to avoid `hub_runtime` if possible. If this is not possible, then maybe there is a lighter alternative to `.init(0)`. There is `hub_runtime.initialize_ports`, but I haven't found suitable arguments for it.

Does anybody know more about this?

**Comment by aivarannamaa** [comment]
I just found out that `Motor("A")` works only when I execute it via REPL. When I put 

```python
import system
from mindstorms import MSHub, Motor, ColorSensor

m = Motor("A")
```

directly to the main.py, then I get `There is not a motor connected to port A`. ColorSensor still doesn't work.

When I leave main.py empty, soft-reboot and submit the same code via REPL, then motor works.

It looks like something important happens between finishing main.py and starting REPL.

**Comment by aivarannamaa** [comment]
Got a little bit closer to my goal with following code

```python
class _Callback:
    def __init__(self, p):
        self.p = p
    def register_persistent(self, x):
        return self.p


import hub
import _api
from util.sensors import initialize_ports as _init_ports

from mindstorms import MSHub, Motor, ColorSensor

_init_ports([
    _Callback(hub.port.A),
    _Callback(hub.port.B),
    _Callback(hub.port.C),
    _Callback(hub.port.D),
    _Callback(hub.port.E),
    _Callback(hub.port.F)
])
```
... but it doesn't avoid those "There is not a ___ connected to port" errors when it is stored in boot.py. Even if it is in main.py, it occasionally doesn't work right.

It only works reliably if I'm calling `_init_ports` in REPL just before constructing a device. It looks like at certain points the effect of `_init_ports` "wears off".

**Comment by aivarannamaa** [comment]
@stefanlyager, perhaps you can tell us how to initialize the sensors and motors for `mindstorms` / `spike` API without starting `hub_runtime`? I'm sorry if it's too invasive, but I'm tagging you because I noticed you were the last one to update https://github.com/LEGO/MINDSTORMS-Robot-Inventor-hub-API and I don't know any other channels for asking such questions from people close to LEGO software.

Is there a place where one can submit feature requests for LEGO software? Do you think LEGO would agree to release the source code of the mpy files on Mindstorms / Spike hubs?

**Reply by gpdaniels** [reply]
I've contacted LEGO's support system about a [reproducible bug I found](https://github.com/gpdaniels/spike-prime/blob/master/simulator/scripts/bug-display.py ) in their hub libraries. Sadly they didn't seem to care about fixing the issue, so I'm not sure they'll take feature requests.

For source code, the first versions of the hub filesystem was not compiled to mpy. I've got dumps of them in this repo:
- https://github.com/gpdaniels/spike-prime/tree/master/filesystem/spike%20-%20v0.5.01.0000-42a938e%20-%200.0.0 
- https://github.com/gpdaniels/spike-prime/tree/master/filesystem/storm%20-%20v0.5.01.0002-f75d82d%20-%200.0.0

Also on the source code front we made some progress to dissassembling and decompiling the mpy files but it's very much not finished. https://github.com/gpdaniels/spike-prime/tree/master/decompiler

**Reply by aivarannamaa** [reply]
Thank you for the response!

> For source code, the first versions of the hub filesystem was not compiled to mpy.

Unfortunately this version has totally different structure, so I can't learn much from it.

> Also on the source code front we made some progress to dissassembling and decompiling the mpy files but it's very much not finished.

I tried it, but couldn't figure out much of the output. 

> Sadly they didn't seem to care about fixing the issue, so I'm not sure they'll take feature requests.

That's a pity. I now decided to try Pybricks API instead.

**Reply by stefanlyager** [reply]
@aivarannamaa

Try something like this:

```python
from system import system
from util.sensors import initialize_ports
from mindstorms import ColorSensor

initialize_ports(system.callbacks.port_callbacks)
cs = ColorSensor("E")
```

I am not aware of a way to submit feature requests. Regarding the release of source code I simply don't know.

**Reply by aivarannamaa** [reply]
@stefanlyager, thanks for the suggestion!

Unfortunately it still gives me error when I replace main.py content with this and soft-reboot. Here's the output after soft-reboot:

```

HUB: sync filesystems
HUB: soft reboot
Traceback (most recent call last):
  File "main.py", line 6, in <module>
  File "_api/colorsensor.py", line 1, in __init__
  File "_api/colorsensor.py", line 1, in _set_mode
AttributeError: 'NoneType' object has no attribute 'mode'
MicroPython v1.14-876-gfbecba865 on 2021-05-04; LEGO Technic Large Hub with STM32F413xx
Type "help()" for more information.
```
